egs/wsj/s5/steps/libs/nnet3/xconfig/basic_layers.py

```diff
@@ -0,0 +1,787 @@
+
+    def get_name(self):
+        """Returns the name of this layer, e.g. 'affine1'.  It does not
+        necessarily correspond to a component name.
+        """
+
+        return self.name
+
+    ######  Functions that might be overridden by the child class: #####
+
+    # child classes should override this.
+    def set_default_configs(self):
+        raise Exception("Child classes must override set_default_configs().")
+
+    # this is expected to be called after set_configs and before check_configs()
+    def set_derived_configs(self):
+        if self.config['dim'] <= 0:
+            self.config['dim'] = self.input_dim()
+
+    # child classes should override this.
+    def check_configs(self):
+        pass
+
+    def get_input_descriptor_names(self):
+        """This function, which may be (but usually will not have to be)
+        overridden by child classes, returns a list of names of the input
+        descriptors expected by this component. Typically this would just
+        return ['input'] as most layers just have one 'input'. However some
+        layers might require more inputs (e.g. cell state of previous LSTM layer
+        in Highway LSTMs). It is used in the function 'normalize_descriptors()'.
+        This implementation will work for layer types whose only
+        Descriptor-valued config is 'input'.
+        If a child class adds more inputs, or does not have an input
+        (e.g. the XconfigInputLayer), it should override this function's
+        implementation to something like: `return ['input', 'input2']`
+        """
+
+        return [ 'input' ]
+
+    def auxiliary_outputs(self):
+        """Returns a list of all auxiliary outputs that this layer supports.
+        These are either 'None' for the regular output, or a string
+        (e.g. 'projection' or 'memory_cell') for any auxiliary outputs that
+        the layer might provide.  Most layer types will not need to override
+        this.
+        """
+
+        return [ None ]
+
+    def output_name(self, auxiliary_output = None):
+        """Called with auxiliary_output == None, this returns the component-node
+        name of the principal output of the layer (or if you prefer, the text
+        form of a descriptor that gives you such an output; such as
+        Append(some_node, some_other_node)).
+        The 'auxiliary_output' argument is a text value that is designed for
+        extensions to layers that have additional auxiliary outputs.
+        For example, to implement a highway LSTM you need the memory-cell of a
+        layer, so you might allow auxiliary_output='memory_cell' for such a
+        layer type, and it would return the component node or a suitable
+        Descriptor: something like 'lstm3.c_t'
+        """
+
+        raise Exception("Child classes must override output_name()")
+
+
+    def input_dim(self):
+        dim = 0
+        for input_name in self.get_input_descriptor_names():
```

Comment:
I don't think this function should exist, at least not with this implementation-- in general we don't want to make any assumptions that if there are multiple input descriptors, they will be appended together.  Having just 1 input is the norm, and auxiliary inputs might not be appended with the primary one [if you wanted that, you could just use Append() in the descriptor.]
Elsewhere the code sets
   input_dim = self.descriptors['input']['dim']
which I think makes more sense [if there is an 'input' descriptor].  If you must have this function, I'd prefer it to use that expression.